Problem: no clear description of the shared values of the project

betatim opened the pull request:
Updated the contributing guide to lay out some ideas about what 
this project should feel like. What to expect in the review process.
What are required pieces to make a PR.

This is not meant as the final word, rather a proposal so please
give feedback. I think it is important to write this down as
some of the concepts like "high quality" can mean many different
things to many different people.

betatim committed:
Updating contribution guide

betatim committed:
Always be rebasing

anaderi commented on CONTRIBUTING.md:
```diff
@@ -35,9 +38,32 @@ To contribute to this repository:
+review. This will send an email to the committers. A valid PR includes
+the following items:
+
+* a description of the problem you are solving
```
maybe just ask to make a reference to specific issue?

anaderi commented on CONTRIBUTING.md:
```diff
@@ -35,9 +38,32 @@ To contribute to this repository:
+* contains the minimal set of changes needed to solve the problem (do
+  not mix several issues, rather create separate PRs)
+* unit and frontend tests for your changes
+* the code should follow the style of the project
```
do we have such defined?

betatim replied:
"pythonic"? Personally not a fan of having something like `pep8` auto run and tell you off because you have 81characters on a line. For me "style" is like a code smell, you can't explain why but some code just smells wrong :-/

anaderi commented on CONTRIBUTING.md:
```diff
@@ -35,9 +38,32 @@ To contribute to this repository:
+* a discussion why you made the choices you made
+* contains the minimal set of changes needed to solve the problem (do
+  not mix several issues, rather create separate PRs)
+* unit and frontend tests for your changes
```
...or justification why tests are not applicable for this PR? (e.g. documentation changes like this one)

anaderi commented:
also a word of caution: http://adamhgrimes.com/blog/the-illusion-of-control/ (i.e. we shouldn't make the contribution process so painful that no one becomes interested, the project should be fun as well, no?)

betatim commented:
Agreed, it should be fun and we should not get stuck in (endless) review cycles.

Was discussing with Gilles how one could combine the best of scikit-learn style with [C4.1](http://rfc.zeromq.org/spec:22). I like the scikit-learn style because it clearly produces excellent and consistent (within itself scikit-learn is very consistent in terms of conventions etc) software. However it is known to take quite some time to get something merged.

Have to re-read C4.1 again to remind myself how they deal with people delivering a valid patch, that isn't "pythonic" for example. As a silly example: `np.max(np_array, axis=1)` vs

```
maxs = []
for i in range(np_array.shape[1]):
  maxs.append(max(np_array[i,:]))
np.array(maxs)
```

both probably pass the tests and do the same thing but "clearly" `np.max()` is the right thing to use. This is what I think of with "style of the project". It is something that is hard to know as an outsider so you need some help to get into it.

betatim committed:
Incorporate comments from @anaderi

betatim commented:
Personally I take the view that that in the long term it is better to not have a feature/more code. Purely from the point of view that maintenance grows exponentially :chart_with_upwards_trend:. You don't want to end up with a code base like LHCb's were no one has a global picture, there is lots of duplication and doing the same thing five different ways. So we need to train up some people who have the global picture in mind and care about keeping it consistent.

If we can solve a problem by contributing upstream: great, now jhub has to maintain it (jk), if we have to add code to everware lets only add good quality code that is easy for us to maintain. Where us is some fuzzy group because we don't know who will be around for how long: :bus:-factor.

I always think of this:

```
Kernighan's lever

Everyone knows that debugging is twice as hard as writing
a program in the first place. So if you're as clever as you can be
when you write it, how will you ever debug it?
```

---

Addressed your comments except for the one on style. Don't have a good answer there. Would be good to hear from other contributors as well.

anaderi commented on CONTRIBUTING.md:
```diff
@@ -35,9 +38,37 @@ To contribute to this repository:
+* the code should follow the style of the project
+* the code should contain documentation in the form of docstrings
+  and comments
+* no commits merging master into your branch, instead rebase your branch
```
in the beginning you write how one can create new branch using git CLI. maybe you can you add a link (like this: https://www.atlassian.com/git/tutorials/merging-vs-rebasing/) to rebase explanation (I'm not sure it's easy to put it here) and sentence of motivation why rebase instead of pull or squash?

betatim replied:
:heavy_check_mark:

anaderi commented on CONTRIBUTING.md:
```diff
@@ -35,9 +38,37 @@ To contribute to this repository:
+apply, write a short sentence in the PR's description explaining why you
+skipped it.
+
+We encourage you to start your PR as soon as possible to get feedback early.
```
the following 3 paragraphs are like soft rules, but the one about 'not merging own PRs' is more hard. maybe put it before softer ones?

betatim replied:
Couldn't think of a nice way to formulate things with changed order so made the final rule bold.

betatim committed:
Link to a rebase howto and motivation for why rebase is preferred

anaderi commented:
LGTM, but since it looks like generic thing, and there is no rush with it, maybe others (@ibab, @astiunov?) also should agree on these changes?

astiunov commented:
I vote for merging

betatim commented:
Rebased and ready for merge.

betatim commented:
:boat: :speedboat:

betatim commented:
Now we just have to live up to the high standards we set for ourselves! I think it'll be worth it even when it feels painful in the moment.
